README
NAME

    String::Util -- String processing utilities

SYNOPSIS

      use String::Util ':all';
    
      # "crunch" whitespace and remove leading/trailing whitespace
      $val = crunch($val);
    
      # does this value have "content", i.e. it's defined
      # and has something besides whitespace?
      if (hascontent $val) {...}
    
      # format for display in a web page
      $val = htmlesc($val);
    
      # format for display in a web page table cell
      $val = cellfill($val);
    
      # remove leading/trailing whitespace
      $val = trim($val);
    
      # ensure defined value
      $val = define($val);
    
      # repeat string x number of times
      $val = repeat($val, $iterations);
    
      # remove leading/trailing quotes
      $val = unquote($val);
    
      # remove all whitespace
      $val = no_space($val);
    
      # remove trailing \r and \n, regardless of what
      # the OS considers an end-of-line
      $val = fullchomp($val);
    
      # or call in void context:
      fullchomp $val;
    
      # encrypt string using random seed
      $val = randcrypt($val);
    
      # are these two values equal, where two undefs count as "equal"?
      if (equndef $a, $b) {...}
    
      # are these two values different, where two undefs count as "equal"?
      if (neundef $a, $b) {...}
    
      # get a random string of some specified length
      $val = randword(10);

DESCRIPTION

    String::Util provides a collection of small, handy utilities for
    processing strings.

INSTALLATION

    String::Util can be installed with the usual routine:

     perl Makefile.PL
     make
     make test
     make install

FUNCTIONS

 crunch(string)

    Crunches all whitespace in the string down to single spaces. Also
    removes all leading and trailing whitespace. Undefined input results in
    undefined output.

 hascontent(scalar), nocontent(scalar)

    hascontent() returns true if the given argument is defined and contains
    something besides whitespace.

    An undefined value returns false. An empty string returns false. A
    value containing nothing but whitespace (spaces, tabs, carriage
    returns, newlines, backspace) returns false. A string containing any
    other characers (including zero) returns true.

    nocontent() returns the negation of hascontent().

 trim(string)

    Returns the string with all leading and trailing whitespace removed.
    Trim on undef returns undef.

    So, for example, the following code changes " my string " to "my
    string":

     $var = " my string  ";
     $var = trim($var);

    trim accepts two optional arguments, 'left' and 'right', both of which
    are true by default. So, to avoid trimming the left side of the string,
    set the 'left' argument to false:

     $var = trim($var, left=>0);

    To avoid trimming the right side, set 'right' to false:

     $var = trim($var, right=>0);

 ltrim, rtrim

    ltrim trims leading whitespace. rtrim trims trailing whitespace. They
    are exactly equivalent to

     trim($var, left=>0);

    and

     trim($var, right=>0);

 no_space(string)

    Removes all whitespace characters from the given string.

 htmlesc(string)

    Formats a string for literal output in HTML. An undefined value is
    returned as an empty string.

    htmlesc is very similar to CGI.pm's escapeHTML. If your script already
    loads CGI.pm, you may well not need htmlesc. However, there are a few
    differences. htmlesc> changes an undefined value to an empty string,
    whereas escapeHTML returns undefs as undefs.

 cellfill(string)

    Formats a string for literal output in an HTML table cell. Works just
    like htmlesc except that strings with no content (i.e. are undef or are
    just whitespace) are returns as &nbsp;.

 jsquote($string)

    Escapes and quotes a string for use in JavaScript. Escapes single
    quotes and surrounds the string in single quotes. Returns the modified
    string.

 unquote(string)

    If the given string starts and ends with quotes, removes them.
    Recognizes single quotes and double quotes. The value must begin and
    end with same type of quotes or nothing is done to the value. Undef
    input results in undef output.

    option: braces

    If the braces option is true, surrounding braces such as [] and {} are
    also removed.

 define(scalar)

    Takes a single value as input. If the value is defined, it is returned
    unchanged. If it is not defined, an empty string is returned.

    This subroutine is useful for printing when an undef should simply be
    represented as an empty string. Granted, Perl already treats undefs as
    empty strings in string context, but this sub makes -w happy. And you
    ARE using -w, right?

 repeat($string, $count)

    Returns the given string repeated the given number of times.

 randword(length, %options)

    Returns a random string of characters. String will not contain any
    vowels (to avoid distracting dirty words). First argument is the length
    of the return string.

    If the string 'dictionary' is sent instead of an integer, then a word
    is randomly selected from a dictionary file. By default, the dictionary
    file is assumed to be at /usr/share/dict/words and the shuf command is
    used to pull out a word. The hash %String::Util::PATHS sets the paths
    to the dictionary file and the shuf executable. Modify that hash to
    change the paths.

    option: alpha

    If the alpha option is true, only alphabetic characters are returned,
    no numerals.

    option: numerals

    If the numerals option is true, only numerals are returned, no
    alphabetic characters.

    option: strip_vowels

    This option is true by default. If true, vowels are not included in the
    returned random string.

 equndef($str1, $str2)

    Returns true if the two given strings are equal. Also returns true if
    both are undef. If only one is undef, or if they are both defined but
    different, returns false.

 neundef($str1, $str2)

    The opposite of equndef, returns true if the two strings are *not* the
    same.

 fullchomp(string)

    Works like chomp, but is a little more thorough about removing \n's and
    \r's even if they aren't part of the OS's standard end-of-line.

    Undefs are returned as undefs.

 randcrypt(string)

    Crypts the given string, seeding the encryption with a random two
    character seed.

 randpost(%opts)

    Returns a string that sorta looks like one or more paragraphs.

    option: word_count

    Sets how many words should be in the post. By default a random number
    from 1 to 250 is used.

    option: par_odds

    Sets the odds of starting a new paragraph after any given word. By
    default the value is .05, which means paragraphs will have an average
    about twenty words.

    option: par

    Sets the string to put at the end or the start and end of a paragraph.
    Defaults to two newlines for the end of a pargraph.

    If this option is a single scalar, that string is added to the end of
    each paragraph.

    To set both the start and end string, use an array reference. The first
    element should be the string to put at the start of a paragraph, the
    second should be the string to put at the end of a paragraph.

    option: max_length

    Sets the maximum length of the returned string, including paragraph
    delimiters.

 ords($string)

    Returns the given string represented as the ascii value of each
    character.

    For example, this code:

     ords('Hendrix')

    returns this string:

     {72}{101}{110}{100}{114}{105}{120}

    options

      * convert_spaces=>[true|false]

      If convert_spaces is true (which is the default) then spaces are
      converted to their matching ord values. So, for example, this code:

       ords('a b', convert_spaces=>1)

      returns this:

      {97}{32}{98}

      This code returns the same thing:

       ords('a b')

      If convert_spaces is false, then spaces are just returned as spaces.
      So this code:

       ords('a b', convert_spaces=>0);

      returns

       {97} {98}

      * alpha_nums

      If the alpha_nums option is false, then characters 0-9, a-z, and A-Z
      are not converted. For example, this code:

       ords('a=b', alpha_nums=>0)

      returns this:

       a{61}b

 deords($string)

    Takes the output from ords() and returns the string that original
    created that output.

    For example, this command:

     deords('{72}{101}{110}{100}{114}{105}{120}')

    returns this string: Hendrix

 crunchlines($str)

    Compacts contiguous newlines into single newlines. Whitespace between
    newlines is ignored, so that two newlines separated by whitespace is
    compacted down to a single newline.

    For example, this code:

     crunchlines("x\n\n\nx")

    outputs two x's with a single empty line between them:

     x
    
     x

TERMS AND CONDITIONS

    Copyright (c) 2012 by Miko O'Sullivan. All rights reserved. This
    program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself. This software comes with NO
    WARRANTY of any kind.

AUTHORS

    Miko O'Sullivan miko@idocs.com

VERSION

    Version 0.10, December 1, 2005

      Initial release

    Version 0.11, December 22, 2005

      This is a non-backwards compatible version.

      urldecode, urlencode were removed entirely. All of the subs that used
      to modify values in place were changed so that they do not do so
      anymore, except for fullchomp.

      See
      http://www.xray.mpe.mpg.de/mailing-lists/modules/2005-12/msg00112.htm
      l for why these changes were made.

    Version 1.01, November 7, 2010

      Decided it was time to upload five years worth of changes.

    Version 1.20, July, 2012

      Properly listing prerequisites.

    Version 1.21, July 18, 2012

      Fixed error in POD. Tightened up code for repet.

    Version 1.22

      Fix in documentation for randpost().

      Clarified documentation for hascontent() and nocontent().

    Version 1.23

      Fixed error in META.yml.

    Version 1.24, December 31, 2014

      Cleaned up POD formatting.

      Changed file to using Unixish style newlines. I hadn't realized until
      now that it was using Windowish newline. How embarrasing.

      Added some features to ords().



Changes
{{$NEXT}}

1.34  2023-02-01

	- Re-release because the required Perl version was wrong

1.33  2023-01-31

	- Remove a bunch of old deprecated functions: crunch, cellfill, define, randword, fullchomp, randcrypt, equndef, neundef

	- Update documentation

1.32  2021-03-26 09:27:35 PDT

	- Per bug reports, trim(undef) will return undef

	- Add a custom separator option to sanitize

1.31  2020-07-27 10:28:07 PDT

	- Deprecate randcrypt() as it's not secure in 2020

	- Add file_get_contents()

	- Make trim() unicode friendly

1.30  2020-07-23 14:29:05 PDT

	- Ownership transferred to Scott Baker

	- Removed undocumented function: spacepad()

	- Added a new sanitize() function

	- Deprecated some older functions: crunch(), fullchomp(), define(), and cellfill()

1.24  2014-12-31

	- Cleaned up POD formatting.

	- Changed file to using Unixish style newlines. I hadn't realized until
	now that it was using Windowish newline. How embarrassing.

	- Added some features to ords().

1.23

	- Fixed error in META.yml.

1.22

	- Fix in documentation for randpost().

	- Clarified documentation for hascontent() and nocontent().

1.21  2012-07-18

	- Fixed error in POD. Tightened up code for repeat.

1.20  2012-07-14

	- Properly listing prerequisites.

1.01  2010-11-07

	- Decided it was time to upload five years' worth of changes.

0.11  2005-12-22

	- This is a non-backwards compatible version.

	- urldecode, urlencode were removed entirely. All of the subs that used
	to modify values in place were changed so that they do not do so
	anymore, except for fullchomp.

	- See
	http://www.xray.mpe.mpg.de/mailing-lists/modules/2005-12/msg00112.html
	for why these changes were made.

0.10  2005-12-01

	- Initial release


test.pl
#!/usr/bin/perl -w
use strict;
use String::Util ':all';
use Test;
BEGIN { plan tests => 11 };

# general purpose variable
my ($val, $org, $new, $got, $should);

# stubs for comparison subroutines
sub err;
sub comp;


#------------------------------------------------------------------------------
# eq_undef, neundef
#
unless (equndef 'a', 'a')
	{ err 'equndef', 'failed comparison of identical defined values' }

unless (equndef undef, undef)
	{ err 'equndef', 'failed comparison of two undefined values' }

if (equndef 'a', 'b')
	{ err 'equndef', 'failed comparison of different defined values' }

if (equndef 'a', undef)
	{ err 'equndef', 'failed comparison of defined and undefined values' }


if (neundef 'a', 'a')
	{ err 'neundef', 'failed comparison of identical defined values' }

if (neundef undef, undef)
	{ err 'neundef', 'failed comparison of two undefined values' }

unless (neundef 'a', 'b')
	{ err 'neundef', 'failed comparison of different defined values' }

unless (neundef 'a', undef)
	{ err 'neundef', 'failed comparison of defined and undefined values' }

ok (1);
#
# eq_undef, neundef
#------------------------------------------------------------------------------



#------------------------------------------------------------------------------
# crunch
#

# basic crunching
$val = "  Starflower \n\n\t  Miko     ";
$val = crunch($val);
comp $val, 'Starflower Miko';

# crunch on undef returns undef
if (defined crunch(undef))
	{ err 'crunch', 'returned defined output for undefined input' }

ok (1);
#
# crunch
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# trim
#

# basic trimming
$val = '  steve     ';
$val = trim($val);
comp $val, 'steve';

# trim on undef returns undef
if (defined trim(undef))
	{ err 'trim', 'returned defined output for undefined input' }

ok (1);
#
# trim
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# define
#

# define an undef
undef $val;
$val = define($val);
comp $val, '';

# define an already defined value
$val = 'x';
define $val;
comp $val, 'x';

ok (1);
#
# define
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# unquote
#

# single quotes
$val = "'Starflower'";
$val = unquote($val);
comp $val, 'Starflower';

# double quotes
$val = '"Starflower"';
$val = unquote($val);
comp $val, 'Starflower';

# no quotes
$val = 'Starflower';
$val = unquote($val);
comp $val, 'Starflower';

ok (1);
#
# unquote
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# htmlesc
#

# basic operation of htmlesc
$val = '<>"&';
$val = htmlesc($val);
comp $val, '&lt;&gt;&quot;&amp;';

# change undef to empty string
undef $val;
$val = htmlesc($val);
comp $val, '';

ok (1);
#
# htmlesc
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# jsquote
#
$val = qq|'yeah\n</script>'|;
$got = jsquote($val);
$should = q|'\'yeah\n<' + '/script>\''|;

comp $got, $should;

ok (1);
#
# jsquote
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# fullchomp
#

# scalar context
$val = qq|Starflower\n\r\r\r\n|;
$val = fullchomp($val);
comp $val, 'Starflower';

ok (1);
#
# fullchomp
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# hascontent
#
undef $val;
hascontent($val) and err 'hascontent', 'returned true on undef';

$val = '';
hascontent($val) and err 'hascontent', 'returned true on empty string';

$val = "   \t   \n\n  \r   \n\n\r     ";
hascontent($val) and err 'hascontent', 'returned true on string that just has whitespace';

$val = '0';
hascontent($val) or err 'hascontent', 'returned false on zero';

$val = ' x ';
hascontent($val) or err 'hascontent', 'returned false defined string with an "x" in it';

ok(1);
#
# hascontent
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# randword
# Not sure how to test this besides making sure it actually runs.
#
undef $val;
$val = randword(20);

unless ( defined($val) && (length($val) == 20) )
	{ err 'randword', 'failed to return random string of appropriate length' }

ok(1);
#
# randword
#------------------------------------------------------------------------------



#------------------------------------------------------------------------------
# randcrypt
# Not sure how to test this besides making sure it actually runs.
#
$val = 'Mypassword';
$val = randcrypt($val);

ok(1);
#
# randcrypt
#------------------------------------------------------------------------------



###############################################################################
# end of tests
###############################################################################


#------------------------------------------------------------------------------
# err
#
sub err {
	my ($function_name, $err) = @_;
	
	print STDERR $function_name, ': ', $err, "\n";
	ok(0);
	exit;
}
#
# err
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# comp
#
sub comp {
	my ($is, $shouldbe) = @_;
	
	if(! equndef($is, $shouldbe)) {
		print STDERR 
			"\n",
			"\tis:         ", (defined($is) ?       $is       : '[undef]'), "\n",
			"\tshould be : ", (defined($shouldbe) ? $shouldbe : '[undef]'), "$shouldbe\n\n";	
		ok(0);
		exit;
	}
}
#
# comp
#------------------------------------------------------------------------------



lib/String/Util.pm
package String::Util;

use strict;
use warnings;
use Carp;
use 5.014;

# version
our $VERSION  = '1.33';
our $FGC_MODE = 'UTF-8';

#------------------------------------------------------------------------------
# opening POD
#

=head1 NAME

B<String::Util> -- String processing utility functions

=head1 DESCRIPTION

B<String::Util> provides a collection of small, handy functions for processing
strings in various ways.

=head1 INSTALLATION

  cpanm String::Util

=head1 USAGE

No functions are exported by default, they must be specified:

  use String::Util qw(trim eqq contains)

alternately you can use C<:all> to export B<all> of the functions

  use String::Util qw(:all)

=head1 FUNCTIONS

=cut

#
# opening POD
#------------------------------------------------------------------------------



#------------------------------------------------------------------------------
# export
#
use base 'Exporter';
use vars qw[@EXPORT_OK %EXPORT_TAGS];

# the following functions accept a value and return a modified version of
# that value
push @EXPORT_OK, qw[
	collapse     htmlesc    trim      ltrim
	rtrim        repeat     unquote   no_space
	nospace      jsquote    crunchlines
	file_get_contents
];

# the following functions return true or false based on their input
push @EXPORT_OK, qw[
	hascontent  nocontent eqq      neqq
	startswith  endswith  contains sanitize
];

# the following function returns the unicode values of a string
push @EXPORT_OK, qw[ ords deords ];

%EXPORT_TAGS = ('all' => [@EXPORT_OK]);
#
# export
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# collapse
#

=head2 collapse($string)

C<collapse()> collapses all whitespace in the string down to single spaces.
Also removes all leading and trailing whitespace.  Undefined input results in
undefined output.

  $var = collapse("  Hello     world!    "); # "Hello world!"

=cut

sub collapse {
	my ($val) = @_;

	if (defined $val) {
		$val =~ s|^\s+||s;
		$val =~ s|\s+$||s;
		$val =~ s|\s+| |sg;
	}

	return $val;
}

#
# collapse
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# hascontent
#

=head2 hascontent($scalar), nocontent($scalar)

C<hascontent()> returns true if the given argument is defined and contains
something besides whitespace.

An undefined value returns false.  An empty string returns false.  A value
containing nothing but whitespace (spaces, tabs, carriage returns, newlines,
backspace) returns false.  A string containing any other characters (including
zero) returns true.

C<nocontent()> returns the negation of C<hascontent()>.

  $var = hascontent("");  # False
  $var = hascontent(" "); # False
  $var = hascontent("a"); # True

  $var = nocontent("");   # True
  $var = nocontent("a");  # False

=cut

sub hascontent {
	my $val = shift();

	if (!defined($val)) {
		return 0;
	}

	# If there are ANY non-space characters in it
	if ($val =~ m|\S|s) {
		return 1;
	}

	return 0;
}

sub nocontent {
	my $str = shift();

	# nocontent() is just the inverse to hascontent()
	my $ret = !(hascontent($str));

	return $ret;
}

#
# hascontent
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# trim
#

=head2 trim($string), ltrim($string), rtrim($string)

Returns the string with all leading and trailing whitespace removed.

  $var = trim(" my string  "); # "my string"

C<ltrim()> trims B<leading> whitespace only.

C<rtrim()> trims B<trailing> whitespace only.

=cut

sub trim {
	my $s = shift();

	if (!defined($s)) {
		return undef;
	}

	$s =~ s/^\s*//u;
	$s =~ s/\s*$//u;

	return $s;
}
#
# trim
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# ltrim, rtrim
#

sub ltrim {
	my $s = shift();

	if (!defined($s)) {
		return undef;
	}

	$s =~ s/^\s*//u;

	return $s;
}

sub rtrim {
	my $s = shift();

	if (!defined($s)) {
		return undef;
	}

	$s =~ s/\s*$//u;

	return $s;
}

#
# ltrim, rtrim
#------------------------------------------------------------------------------



#------------------------------------------------------------------------------
# no_space
#

=head2 nospace($string)

Removes B<all> whitespace characters from the given string. This includes spaces
between words.

  $var = nospace("  Hello World!   "); # "HelloWorld!"

=cut

sub no_space {
	return nospace(@_);
}

# alias nospace to no_space
sub nospace {
	my $val = shift();

	if (defined $val) {
		$val =~ s|\s+||gs;
	}

	return $val;
}

#
# no_space
#------------------------------------------------------------------------------



#------------------------------------------------------------------------------
# htmlesc
#

=head2 htmlesc($string)

Formats a string for literal output in HTML.  An undefined value is returned as
an empty string.

htmlesc() is very similar to CGI.pm's escapeHTML.  However, there are a few
differences. htmlesc() changes an undefined value to an empty string, whereas
escapeHTML() returns undefs as undefs.

=cut

sub htmlesc {
	my ($val) = @_;

	if (defined $val) {
		$val =~ s|\&|&amp;|g;
		$val =~ s|\"|&quot;|g;
		$val =~ s|\<|&lt;|g;
		$val =~ s|\>|&gt;|g;
	} else {
		$val = '';
	}

	return $val;
}
#
# htmlesc
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# jsquote
#

=head2 jsquote($string)

Escapes and quotes a string for use in JavaScript.  Escapes single quotes and
surrounds the string in single quotes.  Returns the modified string.

=cut

sub jsquote {
	my ($str) = @_;

	if (!defined($str)) {
		return undef;
	}

	# Escape single quotes.
	$str =~ s|'|\\'|gs;

	# Break up anything that looks like a closing HTML tag.  This modification
	# is necessary in an HTML web page.  It is unnecessary but harmless if the
	# output is used in a JavaScript document.
	$str =~ s|</|<' + '/|gs;

	# break up newlines
	$str =~ s|\n|\\n|gs;

	# surround in quotes
	$str = qq|'$str'|;

	# return
	return $str;
}
#
# jsquote
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# unquote
#

=head2 unquote($string)

If the given string starts and ends with quotes, removes them. Recognizes
single quotes and double quotes.  The value must begin and end with same type
of quotes or nothing is done to the value. Undef input results in undef output.
Some examples and what they return:

  unquote(q|'Hendrix'|);   # Hendrix
  unquote(q|"Hendrix"|);   # Hendrix
  unquote(q|Hendrix|);     # Hendrix
  unquote(q|"Hendrix'|);   # "Hendrix'
  unquote(q|O'Sullivan|);  # O'Sullivan

B<option:> braces

If the braces option is true, surrounding braces such as [] and {} are also
removed. Some examples:

  unquote(q|[Janis]|, braces=>1);  # Janis
  unquote(q|{Janis}|, braces=>1);  # Janis
  unquote(q|(Janis)|, braces=>1);  # Janis

=cut

sub unquote {
	my ($val, %opts) = @_;

	if (defined $val) {
		my $found = $val =~ s|^\`(.*)\`$|$1|s or
			$val =~ s|^\"(.*)\"$|$1|s or
			$val =~ s|^\'(.*)\'$|$1|s;

		if ($opts{'braces'} && ! $found) {
			$val =~ s|^\[(.*)\]$|$1|s or
			$val =~ s|^\((.*)\)$|$1|s or
			$val =~ s|^\{(.*)\}$|$1|s;
		}
	}

	return $val;
}
#
# unquote
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# repeat
#

=head2 repeat($string, $count)

Returns the given string repeated the given number of times. The following
command outputs "Fred" three times:

  print repeat('Fred', 3), "\n";

Note that C<repeat()> was created a long time based on a misunderstanding of how
the perl operator 'x' works.  The following command using C<x> would perform
exactly the same as the above command.

  print 'Fred' x 3, "\n";

Use whichever you prefer.

=cut

sub repeat {
	my ($val, $count) = @_;
	return ($val x int($count));
}
#
# repeat
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# eqq
# formerly equndef
#

=head2 eqq($scalar1, $scalar2)

Returns true if the two given values are equal.  Also returns true if both
are C<undef>.  If only one is C<undef>, or if they are both defined but different,
returns false. Here are some examples and what they return.

  $var = eqq('x', 'x');     # True
  $var = eqq('x', undef);   # False
  $var = eqq(undef, undef); # True

=cut

sub eqq {
	my ($str1, $str2) = @_;

	# if both defined
	if ( defined($str1) && defined($str2) ) {
		return $str1 eq $str2
	}

	# if neither are defined
	if ( (!defined($str1)) && (!defined($str2)) ) {
		return 1
	}

	# only one is defined, so return false
	return 0;
}
#
# eqq
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# neqq
# formerly neundef
#

=head2 neqq($scalar1, $scalar2)

The opposite of C<neqq>, returns true if the two values are *not* the same.
Here are some examples and what they return.

  $var = neqq('x', 'x');     # False
  $var = neqq('x', undef);   # True
  $var = neqq(undef, undef); # False

=cut

sub neqq {
	return eqq(@_) ? 0 : 1;
}
#
# neqq
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# ords
#

=head2 ords($string)

Returns the given string represented as the ascii value of each character.

  $var = ords('Hendrix'); # {72}{101}{110}{100}{114}{105}{120}

B<options>

=over 4

=item * convert_spaces=>[true|false]

If convert_spaces is true (which is the default) then spaces are converted to
their matching ord values. So, for example, this code:

  $var = ords('a b', convert_spaces=>1); # {97}{32}{98}

This code returns the same thing:

  $var = ords('a b');                    # {97}{32}{98}

If convert_spaces is false, then spaces are just returned as spaces. So this
code:

  ords('a b', convert_spaces=>0);        # {97} {98}


=item * alpha_nums

If the alpha_nums option is false, then characters 0-9, a-z, and A-Z are not
converted. For example, this code:

  $var = ords('a=b', alpha_nums=>0); # a{61}b

=back

=cut

sub ords {
	my ($str, %opts) = @_;
	my (@rv, $show_chars);

	# default options
	%opts = (convert_spaces=>1, alpha_nums=>1, %opts);

	# get $show_chars option
	$show_chars = $opts{'show_chars'};

	# split into individual characters
	@rv = split '', $str;

	# change elements to their unicode numbers
	CHAR_LOOP:
	foreach my $char (@rv) {
		# don't convert space if called so
		if ( (! $opts{'convert_spaces'}) && ($char =~ m|^\s$|s) )
			{ next CHAR_LOOP }

		# don't convert alphanums
		if (! $opts{'alpha_nums'}) {
			if ( $char =~ m|^[a-z0-9]$|si) {
				next CHAR_LOOP;
			}
		}

		my $rv = '{';

		if ($show_chars)
			{ $rv .= $char . ':' }

		$rv .= ord($char) . '}';

		$char = $rv;
	}

	# return words separated by spaces
	return join('', @rv);
}
#
# ords
#------------------------------------------------------------------------------


#------------------------------------------------------------------------------
# deords
#

=head2 deords($string)

Takes the output from C<ords()> and returns the string that original created that
output.

  $var = deords('{72}{101}{110}{100}{114}{105}{120}'); # 'Hendrix'

=cut

sub deords {
	my ($str) = @_;
	my (@tokens, $rv);
	$rv = '';

	# get tokens
	@tokens = split(m|[\{\}]|s, $str);
	@tokens = grep {length($_)} @tokens;

	# build return string
	foreach my $token (@tokens) {
		$rv .= chr($token);
	}

	# return
	return $rv;
}
#
# deords
#------------------------------------------------------------------------------

=head2 contains($string, $substring)

Checks if the string contains substring

  $var = contains("Hello world", "Hello");   # true
  $var = contains("Hello world", "llo wor"); # true
  $var = contains("Hello world", "QQQ");     # false

  # Also works with grep
  @arr = grep { contains("cat") } @input;

=cut

sub contains {
	my ($str, $substr) = @_;

	if (!defined($str)) {
		return undef;
	}

	if (!$substr) {
		$substr = $str;
		$str    = $_;
	}

	my $ret = index($str, $substr, 0) != -1;

	return $ret;
}

=head2 startswith($string, $substring)

Checks if the string starts with the characters in substring

  $var = startwith("Hello world", "Hello"); # true
  $var = startwith("Hello world", "H");     # true
  $var = startwith("Hello world", "Q");     # false

  # Also works with grep
  @arr = grep { startswith("X") } @input;

=cut

sub startswith {
	my ($str, $substr) = @_;

	if (!defined($str)) {
		return undef;
	}

	if (!$substr) {
		$substr = $str;
		$str    = $_;
	}

	my $ret = index($str, $substr, 0) == 0;

	return $ret;
}

=head2 endswith($string, $substring)

Checks if the string ends with the characters in substring

  $var = endswith("Hello world", "world");   # true
  $var = endswith("Hello world", "d");       # true
  $var = endswith("Hello world", "QQQ");     # false

  # Also works with grep
  @arr = grep { endswith("z") } @input;

=cut

sub endswith {
	my ($str, $substr) = @_;

	if (!defined($str)) {
		return undef;
	}

	if (!$substr) {
		$substr = $str;
		$str    = $_;
	}

	my $len   = length($substr);
	my $start = length($str) - $len;

	my $ret = index($str, $substr, $start) != -1;

	return $ret;
}

#------------------------------------------------------------------------------
# crunchlines
#

=head2 crunchlines($string)

Compacts contiguous newlines into single newlines.  Whitespace between newlines
is ignored, so that two newlines separated by whitespace is compacted down to a
single newline.

  $var = crunchlines("x\n\n\nx"); # "x\nx";

=cut

sub crunchlines {
	my ($str) = @_;

	if (!defined($str)) {
		return undef;
	}

	while($str =~ s|\n[ \t]*\n|\n|gs)
		{}

	$str =~ s|^\n||s;
	$str =~ s|\n$||s;

	return $str;
}
#
# crunchlines
#------------------------------------------------------------------------------

=head2 sanitize($string, $separator = "_")

Sanitize all non alpha-numeric characters in a string to underscores.
This is useful to take a URL, or filename, or text description and know
you can use it safely in a URL or a filename.

B<Note:> This will remove any trailing or leading '_' on the string

  $var = sanitize("http://www.google.com/") # http_www_google_com
  $var = sanitize("foo_bar()";              # foo_bar
  $var = sanitize("/path/to/file.txt");     # path_to_file_txt
  $var = sanitize("Big yellow bird!", "."); # Big.yellow.bird

=cut

sub sanitize {
    my $str = shift();
    my $sep = shift() // "_";

    if (!defined($str)) {
        return undef;
    }

    # Convert multiple non-word sequences to the separator
    $str =~ s/[\W_]+/$sep/g;

    # The separator is a literal character so we quotemeta it
    $sep = quotemeta($sep);
    # Remove any separators at the beginning and end
    $str =~ s/\A$sep+//;
    $str =~ s/$sep+\z//;

    return $str;
}

###########################################################################

=head2 file_get_contents($string, $boolean)

Read an entire file from disk into a string. Returns undef if the file
cannot be read for any reason. Can also return the file as an array of
lines.

  $str   = file_get_contents("/tmp/file.txt");    # Return a string
  @lines = file_get_contents("/tmp/file.txt", 1); # Return an array

B<Note:> If you opt to return an array, carriage returns and line feeds are
removed from the end of each line.

B<Note:> File is read in B<UTF-8> mode, unless C<$FGC_MODE> is set to an
appropriate encoding.

=cut

sub file_get_contents {
	my ($file, $ret_array) = @_;

	open (my $fh, "<", $file) or return undef;
	binmode($fh, ":encoding($FGC_MODE)");

	if ($ret_array) {
		my @ret;

		while (my $line = readline($fh)) {
			$line =~ s/[\r\n]*$//; # Remove CR/LF
			push(@ret, $line);
		}

		return @ret;
	} else {
		my $ret = '';
		while (my $line = readline($fh)) {
			$ret .= $line;
		}

		return $ret;
	}
}

# return true
1;


__END__

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012-2016 by Miko O'Sullivan.  All rights reserved.  This program
is free software; you can redistribute it and/or modify it under the same terms
as Perl itself. This software comes with B<NO WARRANTY> of any kind.

=head1 AUTHORS

Miko O'Sullivan <miko@idocs.com>

Scott Baker <scott@perturb.org>

=cut



#------------------------------------------------------------------------------
# module info
# This info is used by a home-grown CPAN builder. Please leave it as it is.
#
{
	// include in CPAN distribution
	include : 1,

	// test scripts
	test_scripts : {
		'Util/tests/test.pl' : 1,
	},
}
#
# module info
#------------------------------------------------------------------------------
